### PR TITLE
Make pallet macro generate accessor to PalletInfo information on pallet placeholder

### PR DESCRIPTION
**Pull request by thiolliere**
https://github.com/paritytech/substrate/issues/8583

`PalletInfoAccess` trait is introduced, and implemented by pallet macro on `Pallet`.

### Controversial

* Maybe we can make it more explicit by only generating them when user give the attribute: `#[implement_pallet_info_access)]` or something similar.
* (We can also implement `fn index` and `fn name` directly on the struct to avoid having to import the trait in scope, but this makes generation less expectable and can break user code if they implement a function with such a name already)

### CODE REVIEW

**Comment by bkchr** [comment]
Why not make it a trait?

**Comment by thiolliere** [comment]
> Why not make it a trait?

I thought maybe ppl would be annoyed to have to import the trait in scope.
But yes a trait seems better indeed, I'm doing so

**Comment by thiolliere** [comment]
now implemented with trait.

**Comment by bkchr** [comment]
We could put the trait into the prelude

File: frame/support/src/traits/metadata.rs [hunk 1]
```diff
@@ -31,6 +31,16 @@ pub trait PalletInfo {
 	fn name<P: 'static>() -> Option<&'static str>;
 }
 
+/// Provides information about the pallet setup in the runtime.
+///
+/// Access the information provided by [`PalletInfo`] for a specific pallet.
+pub trait PalletInfoAccess {
```
Comment (shawntabrizi) [inline frame/support/src/traits/metadata.rs:37]: Is `PalletInfo` already taken? `Access` is weird.

**Review by shawntabrizi — APPROVED** [review]

File: frame/support/src/traits/metadata.rs (hunk 1 above)
Comment (thiolliere) [reply frame/support/src/traits/metadata.rs:37]: `PalletInfo` is the trait giving the info for all pallets https://substrate.dev/rustdocs/v3.0.0/frame_support/traits/trait.PalletInfo.html
It is implemented by an associated type on frame-system https://substrate.dev/rustdocs/v3.0.0/frame_system/pallet/trait.Config.html#associatedtype.PalletInfo

Ideally `PalletInfo` could be renamed `PalletsInfo` as it contains info for all pallets, and thus we could use `PalletInfo` here.

Anyway those 2 traits are very related, `PalletInfoAccess` is implementing accessor to the info of the pallet. info of the pallet are retrieved from `PalletInfo` which contains info for all pallets (but which is cumbersome to use directly).

**Comment by thiolliere** [comment]
bot merge

File: frame/support/src/traits/metadata.rs (hunk 1 above)
Comment (kianenigma) [reply frame/support/src/traits/metadata.rs:37]: > Ideally PalletInfo could be renamed PalletsInfo as it contains info for all pallets, and thus we could use PalletInfo here.

Totally agree with this, maybe even in this PR? the naming is weird, but the functionality is great, working with the old `PalletInfo` was not simple.

**Review by kianenigma — APPROVED** [review]

File: frame/support/src/traits/metadata.rs (hunk 1 above)
Comment (thiolliere) [reply frame/support/src/traits/metadata.rs:37]: At the same time I prefer not to break, as `PalletInfoAccess` is also an OK name to me.
Maybe we can write it somewhere for when we want to break frame stuff.

**Comment by thiolliere** [comment]
bot merge

File: frame/support/src/traits/metadata.rs (hunk 1 above)
Comment (kianenigma) [reply frame/support/src/traits/metadata.rs:37]: sure, lets stay backwards compat.

**Pull request merged**